src/qutip_cupy/__init__.py

```diff
@@ -36,10 +36,22 @@
 )
 data.to.register_aliases(["cupyd"], CuPyDense)
 
-data.adjoint.add_specialisations([(CuPyDense, CuPyDense, cd.cpd_adjoint)])
-data.transpose.add_specialisations([(CuPyDense, CuPyDense, cd.cpd_transpose)])
-data.conj.add_specialisations([(CuPyDense, CuPyDense, cd.cpd_conj)])
-data.trace.add_specialisations([(CuPyDense, cd.cpd_trace)])
+data.adjoint.add_specialisations([(CuPyDense, CuPyDense, cd.adjoint_cupydense)])
+data.transpose.add_specialisations([(CuPyDense, CuPyDense, cd.transpose_cupydense)])
+data.conj.add_specialisations([(CuPyDense, CuPyDense, cd.conj_cupydense)])
+data.trace.add_specialisations([(CuPyDense, cd.trace_cupydense)])
+data.mul.add_specialisations([(CuPyDense, CuPyDense, cd.mul_cupydense)])
+data.imul.add_specialisations([(CuPyDense, CuPyDense, cd.imul_cupydense)])
+data.neg.add_specialisations([(CuPyDense, CuPyDense, cd.neg_cupydense)])
+data.matmul.add_specialisations(
+    [(CuPyDense, CuPyDense, CuPyDense, cd.matmul_cupydense)]
+)
+data.add.add_specialisations([(CuPyDense, CuPyDense, CuPyDense, cd.add_cupydense)])
+data.sub.add_specialisations([(CuPyDense, CuPyDense, CuPyDense, cd.sub_cupydense)])
+
+data.diag.add_specialisations([(CuPyDense, cd.diags)])
```

Comment:
I made the same question on issue #15 see [this](f.append(experiment_dict) I am sticking to the naming used in qutip dev.major, but I think the proper thing would be to extend the naming convention as I suggest there,

Reply:
In fact `diags` is the name given in `dev.major`  but I am happier with `diag`.

Reply:
Although if we were to apply the same principle as on PR #15 , we should keep`diags` because a later developer might be confused by the name chages between `qutip`  and `qutip-cupy`

Reply:
It's `diags` rather than `diag` because that matched `qutip.qdiags`, and it can accept a list of diagonals rather than just one (as opposed to `np.diag`).  I was mostly just matching pre-existing stuff in QuTiP, but in general I think `diags` makes sense.  I commented on the historical reasons for the naming convention somewhere else, but the gist was that originally I assumed that anything defined in `csr.pyx` or `dense.pyx` would only ever be accessed by qualified access through that module, and having the additional type information was superfluous, consequently adding noise to the code when you're reading it.  It would be clear, say, that you needed a type suffix on a `norm.l2_*` function, though, because those don't have the type in the module name.

From the dispatcher user's perspective, it really doesn't matter what they're called, because they never see the raw objects anyway.  I'm not against renaming things like `csr.diags` to `diags_csr`, and having data's `__init__.pxd` and `__init__.py` import those select names from `CSR`.